object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java

```diff
@@ -0,0 +1,515 @@
+      BaseTypeChecker checker,
+      CFAnalysis analysis) {
+    this.typeFactory = typeFactory;
+    this.checker = checker;
+    this.analysis = analysis;
+  }
+
+  /**
+   * This function traverses the given method CFG and reports an error if "f" isn't called on any
+   * local variable node whose class type has @MustCall(f) annotation before the variable goes out
+   * of scope. The traverse is a standard worklist algorithm. Worklist and visited entries are
+   * BlockWithLocals objects that contain a set of (LocalVariableNode, Tree) pairs for each block. A
+   * pair (n, T) represents a local variable node "n" and the latest AssignmentTree "T" that assigns
+   * a value to "n".
+   *
+   * @param cfg the control flow graph of a method
+   */
+  void checkMustCallInvoked(ControlFlowGraph cfg) {
+    // add any owning parameters to initial set of variables to track
+    BlockWithLocals firstBlockLocals =
+        new BlockWithLocals(cfg.getEntryBlock(), computeOwningParameters(cfg));
+
+    Set<BlockWithLocals> visited = new HashSet<>();
+    Deque<BlockWithLocals> worklist = new ArrayDeque<>();
+
+    worklist.add(firstBlockLocals);
+    visited.add(firstBlockLocals);
+
+    while (!worklist.isEmpty()) {
+
+      BlockWithLocals curBlockLocals = worklist.removeLast();
+      List<Node> nodes = curBlockLocals.block.getNodes();
+      // defs to be tracked in successor blocks, updated by code below
+      Set<LocalVarWithTree> newDefs = new HashSet<>(curBlockLocals.localSetInfo);
+
+      for (Node node : nodes) {
+
+        if (node instanceof AssignmentNode) {
+          handleAssignment((AssignmentNode) node, newDefs);
+        } else if (node instanceof ReturnNode) {
+          handleReturn((ReturnNode) node, cfg, newDefs);
+        } else if (node instanceof MethodInvocationNode || node instanceof ObjectCreationNode) {
+          handleInvocation(newDefs, node);
+        }
+      }
+
+      handleSuccessorBlocks(visited, worklist, newDefs, curBlockLocals.block);
+    }
+  }
+
+  private void handleInvocation(Set<LocalVarWithTree> newDefs, Node node) {
+    List<Node> arguments;
+    ExecutableElement executableElement;
+    if (node instanceof MethodInvocationNode) {
+      MethodInvocationNode invocationNode = (MethodInvocationNode) node;
+      arguments = invocationNode.getArguments();
+      executableElement = TreeUtils.elementFromUse(invocationNode.getTree());
+    } else {
+      arguments = ((ObjectCreationNode) node).getArguments();
+      executableElement = TreeUtils.elementFromUse(((ObjectCreationNode) node).getTree());
+    }
+
+    List<? extends VariableElement> formals = executableElement.getParameters();
+    if (arguments.size() != formals.size()) {
+      // this could happen, e.g., with varargs, or with strange cases like generated Enum
+      // constructors
+      // for now, just skip this case
+      // TODO allow for ownership transfer here if needed in future
+      return;
```

Comment:
We don't have to address this now, but it seems to indicate that there's currently a soundness hole in our ownership transfer algorithm if a varargs parameter is marked as `@Owning`. We should probably add validation that there isn't such an `@Owning` annotation and issue a warning if there is in the visitor.

Reply:
I don't think this is a soundness hole.  We will just fail to transfer ownership to the varargs formal parameter, which means we will get an error somewhere else if the `@MustCall` obligation is not satisfied.  I agree we could add a warning if someone writes `@Owning` on a varargs parameter to make this more evident to the user.  I will add to my TODO list